### PR TITLE
MINOR add GH action for PR builds

### PR DESCRIPTION
**Pull request by mumrah**
This PR adds a GitHub Action for running the junit tests on PRs.

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  validate:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        java: [8, 21]
+    name: Compile and Check Java ${{ matrix.java }}
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
```
Comment (chia7712) [inline .github/workflows/pr.yml]: Should we use `v4` instead? the action running for docker is using `v4`

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      with:
+        persist-credentials: false
+    - name: Setup Java
+      uses: actions/setup-java@v4
+      with:
+        distribution: temurin
+        java-version: ${{ matrix.java }}
+    - name: Setup Gradle
+      uses: gradle/actions/setup-gradle@v3
```
Comment (chia7712) [inline .github/workflows/pr.yml]: ditto. maybe we can use `v4`?

File: .github/workflows/pr.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    - name: Compile and validate
+      run: ./gradlew --build-cache --info --scan check -x test
+
+  test:
+    needs: validate
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        java: [ 8, 21 ]
```
Comment (chia7712) [inline .github/workflows/pr.yml]: in jenkins env, we run check on jdk 8 and 21, and run tests on jdk 11 and 17. Maybe we should keep consistency?

File: .github/workflows/pr.yml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        uses: actions/checkout@v3
+        with:
+          persist-credentials: false
+      - name: Setup Java
+        uses: actions/setup-java@v4
+        with:
+          distribution: temurin
+          java-version: ${{ matrix.java }}
+      - name: Setup Gradle
+        uses: gradle/actions/setup-gradle@v3
```
Comment (chia7712) [inline .github/workflows/pr.yml]: ditto. maybe `v4`?

**Review by chia7712 — COMMENTED** [review]
@mumrah thanks for this nice patch! a couple of comments are left. Also, why the Github action does not run on kafka repo? I can see the action run on your repo (https://github.com/mumrah/kafka/actions/runs/10422753706/job/28867954590)

**Comment by mumrah** [comment]
Thanks for taking a look @chia7712! 

> Also, why the Github action does not run on kafka repo?

I believe the workflow has to exist on trunk first. Once it's there PRs should be trigger the workflow. PRs can also alter workflow and see the new results (useful for testing CI changes before committing). We will probably want to only allow committers to trigger actions with modified workflow files.

I'll check with Infra about the policy for committers vs non-committers triggering workflow runs. There are some org-level settings in GitHub related to this.

File: .github/workflows/pr.yml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    types: [opened, synchronize, reopened]
+    branches: 'gh-*'
+
+jobs:
+  validate:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        java: [8, 11, 17, 21]
```
Comment (chia7712) [inline .github/workflows/pr.yml:29]: I'm ok to see more checks. or it is fine to keep jdk8 and jdk21 as jenkins CI does

File: .github/workflows/pr.yml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      uses: actions/setup-java@v4
+      with:
+        distribution: temurin
+        java-version: ${{ matrix.java }}
+    - name: Setup Gradle
+      uses: gradle/actions/setup-gradle@v4
+      env:
+        GRADLE_BUILD_ACTION_CACHE_DEBUG_ENABLED: true
+      with:
+        gradle-version: current
```
Comment (chia7712) [inline .github/workflows/pr.yml]: the patch #16908 used to update gradle from 8.8 to 8.10 (current) is not merged. Hence, it is a little tricky to decide the version here. If we keep using `current`, CI will have useless 8.10 and then download 8.8 later. 

it is not a big issue to adopt `current` as we will use "current" version eventually.

**Review by chia7712 — APPROVED** [review]
@mumrah thanks for this patch. LGTM two non-blocker comments are left. PTAL

File: .github/workflows/pr.yml (hunk 5 above)
Comment (mumrah) [reply .github/workflows/pr.yml:29]: Let's change this in a future PR so we can test altering the workflows in a PR :)

**Pull request merged**